### PR TITLE
Py2: Deterministic install/upgrade order for eopkg & baselayout

### PR DESCRIPTION
**Pull request by ermo**
This will ensure that future upgrades to eopkg and baselayout will always happen in a (semi-)deterministic fashion.

- If eopkg is part of the install or upgrade transaction, eopkg will be moved to the front of the transaction.
- If baselayout is part of the install or upgrade transaction, baselayout will be moved to the front of the transaction (possibly in front of eopkg).

For context, only the ordering between eopkg, baselayout and any system.base packages which are part of the transaction is currently preserved.

The install order in the presence of system.base packages can potentially also be tweaked now that we're using ordered sets where it matters (if people are not happy with the current logic).

Note that this commit requires you to build and install the [python2-ordered-set PR](https://github.com/getsolus/packages/pull/3057) that @sheepman4267 prepared.

**How to use this branch as a daily driver for testing**

- (git clone the eopkg repo somewhere) / `git pull -a`
- `git checkout py2-eopkg-and-baselayout-install-upgrade-order`
- Add the following function (or its equivalent) to your shell startup scripts:
```bash
function eopkg () {
    pushd ~/repos/getsolus/pisi-py2/ # replace with your own location!
    # if the invocation does not exit 0, still return to the previous location
    sudo ./eopkg.py2 --debug ${*} || :
    popd
}
```

**Test case**

This test case consistently reinstalls baselayout, pisi and eopkg first (in that order). Any other packages may arrive in whatever order the topological sort used by pisi determines is valid (it is not stable).

```
ermo@solbox:~/repos/getsolus/pisi-py2 [py2-eopkg-and-baselayout-install-upgrade-order* +1 ~0 -0 !]
$ sudo ./eopkg.py2 --debug it --reinstall baselayout eopkg pisi htop solbuild nano
DEBUG: InstallDB initialized in 0.00686407089233.
DEBUG: PackageDB initialized in 0.0629091262817.
DEBUG: RepoDB initialized in 0.000216007232666.
DEBUG: HistoryDB initialized in 0.0993731021881.
DEBUG: ComponentDB initialized in 0.00314283370972.
DEBUG: RepoDB initialized in 7.29560852051e-05.
DEBUG: InstallDB initialized in 0.00521898269653.
DEBUG: PackageDB initialized in 0.0601711273193.
digraph G {


}
DEBUG: checking glibc release >= 114
DEBUG: checking python
DEBUG: checking python2-xattr
DEBUG: checking db5
DEBUG: checking piksemel
DEBUG: checking usysconf
DEBUG: checking ncurses release >= 21
DEBUG: checking libcap2 release >= 15
DEBUG: checking glibc release >= 98
DEBUG: checking lm_sensors
DEBUG: checking glibc release >= 114
DEBUG: checking git
DEBUG: checking file release >= 24
DEBUG: checking glibc release >= 110
DEBUG: checking ncurses release >= 25
DEBUG: checking nanorc
digraph G {
pisi[ label = "pisi(3.11,111)" ];
nano[ label = "nano(8.0,195)" ];
eopkg[ label = "eopkg(4.0.0,6)" ];
htop[ label = "htop(3.3.0,22)" ];
solbuild[ label = "solbuild(1.6.3,54)" ];
baselayout[ label = "baselayout(1.9.0,76)" ];


}
topological_sort order: ['baselayout', 'solbuild', 'htop', 'eopkg', 'nano', 'pisi']
deterministic order: ['solbuild', 'htop', 'nano', 'pisi', 'eopkg', 'baselayout']
final order.reverse(): ['baselayout', 'eopkg', 'pisi', 'nano', 'htop', 'solbuild']
Reordering install order so system.base packages come first.
install_order: ['baselayout', 'pisi', 'eopkg', 'nano', 'htop', 'solbuild']
Following packages will be installed:
baselayout  eopkg  htop  nano  pisi  solbuild
Total size of package(s): 22.10 MB
Downloading 1 / 6
Package baselayout found in repository Unstable
baselayout-1.9.0-76-1-x86_64.eopkg [cached]
Downloading 2 / 6
Package pisi found in repository Unstable
pisi-3.11-111-1-x86_64.eopkg [cached]
Downloading 3 / 6
Package eopkg found in repository Unstable
eopkg-4.0.0-6-1-x86_64.eopkg [cached]
Downloading 4 / 6
Package nano found in repository Unstable
nano-8.0-195-1-x86_64.eopkg [cached]
Downloading 5 / 6
Package htop found in repository Unstable
htop-3.3.0-22-1-x86_64.eopkg [cached]
Downloading 6 / 6
Package solbuild found in repository Unstable
solbuild-1.6.3-54-1-x86_64.eopkg [cached]
Installing 1 / 6
baselayout-1.9.0-76-1-x86_64.eopkg [cached]
Installing baselayout, version 1.9.0, release 76
DEBUG: FilesDB initialized in 0.00332593917847.
Extracting the files of baselayout
Installed baselayout
Installing 2 / 6
pisi-3.11-111-1-x86_64.eopkg [cached]
Installing pisi, version 3.11, release 111
Extracting the files of pisi
Installed pisi
Installing 3 / 6
eopkg-4.0.0-6-1-x86_64.eopkg [cached]
Installing eopkg, version 4.0.0, release 6
Extracting the files of eopkg
Installed eopkg
Installing 4 / 6
nano-8.0-195-1-x86_64.eopkg [cached]
Installing nano, version 8.0, release 195
Extracting the files of nano
Installed nano
Installing 5 / 6
htop-3.3.0-22-1-x86_64.eopkg [cached]
Installing htop, version 3.3.0, release 22
Extracting the files of htop
Installed htop
Installing 6 / 6
solbuild-1.6.3-54-1-x86_64.eopkg [cached]
Installing solbuild, version 1.6.3, release 54
Extracting the files of solbuild
Installed solbuild
 [✓] Syncing filesystems                                                success
 [✓] Updating dynamic library cache                                     success
 [✓] Updating systemd tmpfiles                                          success
 [✓] Updating icon theme cache: hicolor                                 success
 [✓] Updating desktop database                                          success
 [✓] Updating manpages database                                         success
ermo@solbox:~/repos/getsolus/pisi-py2 [py2-eopkg-and-baselayout-install-upgrade-order* +1 ~0 -0 !]
$ 
```

### CODE REVIEW

**Review by silkeh — CHANGES_REQUESTED** [review]
Few minor comments. Looks simple enough that it can't really break.

I tested it in a VM just to be sure, and looks like update works. Couldn't test history rollback though, as this VM was too old :see_no_evil:. This change looks fairly safe to me, so I don't think it needs extended (weeks/months long) testing.

Will use this as a daily driver on my notebook when I use it next.

File: pisi/operations/helper.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 import os
 
+from ordered_set import OrderedSet as set
```
[the remark this answers is not in the document]
Comment (ermo) [reply pisi/operations/helper.py:15]: If I'm honest, I'd rather split out stuff where versions differ between the py2-compat stuff to explicit `python2-<somename>` and then keep the existing `python-<somename>` as pure py3 packages?

@sheepman4267 has this PR up: https://github.com/getsolus/packages/pull/3057

File: pisi/operations/install.py [hunk 2]
```diff
@@ -25,6 +26,38 @@
...
+        after = order[pkg_index+1:]
+        new_order = before + after
+        new_order.append(pkg)
+        return new_order
+
+    # when eopkg is in the order, move it to the end of the order (since the order gets reversed).
+    # this is useful when file ownership between it and pisi change.
+    if EOPKG_PKG in order:
+        order = place_at_end(EOPKG_PKG, order)
+        #return [EOPKG_PKG]
```
[the remark this answers is not in the document]
Comment (ermo) [reply pisi/operations/install.py:54]: This is here if we decide that, actually, anything that has an eopkg update should _only_ do that update first.

I left it in to remind everyone that this is (also) an option.

File: pisi/operations/install.py [hunk 3]
```diff
@@ -25,6 +26,38 @@
...
+    def place_at_end(pkg, order):
+        """place pkg at the end of the ordered list 'order'"""
+        #print("pkg is of type: %s, order is of type: %s" % (type(pkg), type(order)))
+        order = list(order)
+        pkg_index = order.index(pkg)
+        before = order[:pkg_index]
+        after = order[pkg_index+1:]
+        new_order = before + after
+        new_order.append(pkg)
+        return new_order
```
[the remark this answers is not in the document]
Comment (ermo) [reply pisi/operations/install.py:48]: `order` is an ordered set already.

So maybe the best implementation is to remove the function entirely and just use a sequence of:

    order.remove(pkg)
    order.add(pkg

File: pisi/operations/install.py [hunk 4]
```diff
@@ -25,6 +26,38 @@
...
+EOPKG_PKG = 'eopkg'
+
+def plan_deterministic_install_order(order):
+    """Take an install order and ensure that it is sane."""
+
+    # save cycles
+    if len(order) <= 1:
+        return order
+
+    def place_at_end(pkg, order):
```
[the remark this answers is not in the document]
Comment (ermo) [reply pisi/operations/install.py:39]: See comment below.

File: pisi/operations/install.py [hunk 5]
```diff
@@ -25,6 +26,38 @@
...
+        """place pkg at the end of the ordered list 'order'"""
+        #print("pkg is of type: %s, order is of type: %s" % (type(pkg), type(order)))
```
[the remark this answers is not in the document]
Comment (ermo) [reply pisi/operations/install.py:41]: And a useful one too.

**Review by silkeh — APPROVED** [review]
LGTM! What could go wrong?

**Pull request merged**